unittests/db_to_kv_tests.cpp

```diff
@@ -0,0 +1,309 @@
+
+   BOOST_CHECK(key == decomposed_key);
+   BOOST_CHECK_EQUAL(primary_key, decomposed_primary_key);
+
+   verify_secondary_wont_convert(composite_key, key_type::sec_i256);
+
+   std::vector<eosio::chain::key256_t> keys;
+   keys.push_back(create_key256(create_uint128(0x0, 0x0), create_uint128(0x0, 0x0)));
+   keys.push_back(create_key256(create_uint128(0x0, 0x0), create_uint128(0x0, 0x01)));
+   keys.push_back(create_key256(create_uint128(0x0, 0x0), create_uint128(0x0, 0x7FFF'FFFF'FFFF'FFFF)));
+   keys.push_back(create_key256(create_uint128(0x0, 0x0), create_uint128(0x0, 0x8000'0000'0000'0000)));
+   keys.push_back(create_key256(create_uint128(0x0, 0x0), create_uint128(0x0, 0xFFFF'FFFF'FFFF'FFFF)));
+   keys.push_back(create_key256(create_uint128(0x0, 0x0), create_uint128(0x1, 0x0)));
+   keys.push_back(create_key256(create_uint128(0x0, 0x0), create_uint128(0xFFFF'FFFF'FFFF'FFFF, 0x0)));
+   keys.push_back(create_key256(create_uint128(0x0, 0x1), create_uint128(0x0, 0x0)));
+   keys.push_back(create_key256(create_uint128(0x0, 0xFFFF'FFFF'FFFF'FFFF), create_uint128(0x0, 0x0)));
+   keys.push_back(create_key256(create_uint128(0x1, 0x0), create_uint128(0x0, 0x0)));
+   keys.push_back(create_key256(create_uint128(0xFFFF'FFFF'FFFF'FFFF, 0x0), create_uint128(0x0, 0x0)));
+   check_ordered_keys(keys, &eosio::chain::db_key_value_format::create_secondary_key<eosio::chain::key256_t>);
+}
+
+BOOST_AUTO_TEST_CASE(float64_key_conversions_test) {
+   const name scope = N(myscope);
+   const name table = N(thisscope);
+   const float64_t key = to_softfloat64(6.0);
+   const uint64_t primary_key = 0x0123456789abcdef;
+   const auto composite_key = eosio::chain::db_key_value_format::create_secondary_key(scope, table, key, primary_key);
+   BOOST_REQUIRE_EQUAL(overhead_size + sizeof(key) + sizeof(primary_key), composite_key.size());
+
+   name decomposed_scope;
+   name decomposed_table;
+   float64_t decomposed_key = to_softfloat64(0.0);
+   uint64_t decomposed_primary_key = 0;
+   BOOST_REQUIRE_NO_THROW(eosio::chain::db_key_value_format::get_secondary_key(composite_key, decomposed_scope, decomposed_table, decomposed_key, decomposed_primary_key));
+   BOOST_CHECK_EQUAL(scope.to_string(), decomposed_scope.to_string());
+   BOOST_CHECK_EQUAL(table.to_string(), decomposed_table.to_string());
+
+   BOOST_CHECK_EQUAL(key, decomposed_key);
+   BOOST_CHECK_EQUAL(primary_key, decomposed_primary_key);
+
+   verify_secondary_wont_convert(composite_key, key_type::sec_double);
+
+   std::vector<float64_t> keys;
+   keys.push_back(to_softfloat64(-100000.0));
+   keys.push_back(to_softfloat64(-99999.9));
+   keys.push_back(to_softfloat64(-1.9));
+   keys.push_back(to_softfloat64(0.0));
+   keys.push_back(to_softfloat64(0.1));
+   keys.push_back(to_softfloat64(1.9));
+   keys.push_back(to_softfloat64(99999.9));
+   keys.push_back(to_softfloat64(100000.0));
+   check_ordered_keys(keys, &eosio::chain::db_key_value_format::create_secondary_key<float64_t>);
```

Comment:
Do we need to check for cases with denormalized floating point values and other things like NaN's?

Reply:
IIRC, the db intrinsics forbid NaN keys.  Subnormal and infinities are both valid, however.

Reply:
NaNs are precluded, added test for -inf and inf ordering compared to min and max.  What others do we need to check for? @b1bart